### ISSUE TITLE
Improve handling of indexing keywords `begin` and `end` when using `@which` macro

### ISSUE THREAD

**Issue by bernb** [JuliaLang/julia#37379]
I understand there are certain limitations to macros:
```
julia> @which rand(10) .* rand(10)
ERROR: dot expressions are not lowered to a single function call, so @which cannot analyze them. You may want to use Meta.@lower to identify which function call to target.
```
This gives an explanation and advice. The situation for `begin` and `end` is worse, unfortunately:
```
julia> a = [1,2]; @which a[1:end]
ERROR: UndefVarError: end not defined
```
From my understanding, `begin` and `end` in this context get simply translated to `firstindex` and `lastindex`, so as an user an error is a bit suprising. I believe a more expressive error message would be helpful as long as `@which` can't resolve those keywords.

**Comment by rapus95** [comment]
The same UndefVarError occurs for all `@code_*` macros which to me looks like a flaw.

**Comment by sostock** [comment]
The `begin`/`end` issue should be easy to fix with `Base.replace_ref_begin_end!` as used by the `@view` macro.

**Comment by ViralBShah** [comment]
Extensive discussion in #29920

**Comment by giordano** [comment]
I don't think #29920 is related to this issue.

**Comment by ianatol** [comment]
This was fixed by #45968
